src/SFML/Audio/AudioResource.cpp

```diff
@@ -28,34 +28,45 @@
 #include <SFML/Audio/AudioDevice.hpp>
 #include <SFML/Audio/AudioResource.hpp>
 
-#include <memory>
 #include <mutex>
 
 
+namespace
+{
+
+std::mutex             deviceMutex;
+sf::priv::AudioDevice* device;
+unsigned int           deviceRC{};
+
+} // namespace
+
 namespace sf
 {
 ////////////////////////////////////////////////////////////
-AudioResource::AudioResource() :
-m_device(
-    []
-    {
-        // Ensure we only ever create a single instance of an
-        // AudioDevice that is shared between all AudioResources
-        static std::mutex                       mutex;
-        static std::weak_ptr<priv::AudioDevice> weakAudioDevice;
+AudioResource::AudioResource()
+{
+    const std::lock_guard guard{deviceMutex};
+
+    if (deviceRC++ == 0u)
+        device = new sf::priv::AudioDevice;
```

Comment:
I'm very much surprised that you'd suggest a custom reference counting implementation in a C++17 code base, especially while actively replacing a shared/weak_ptr implementation. To what end? Save a std header include.

This goes too far for me. I don't want this in SFML, nor would I ever recommend to anyone to do this.

Reply:
Thank you for your feedback. I'd like to clarify my rationale behind suggesting a custom reference counting implementation for this specific use case.

Our requirements are straightforward:

- Maintain a single `sf::priv::AudioDevice` global object for the entire program.

- Ensure the first `AudioResource` object initializes the audio device if none are active.

- De-initialize the audio device when the last `AudioResource` object is destroyed.

While `std::shared_ptr` and `std::weak_ptr` are powerful tools, they offer functionalities that exceed our needs, such as support for atomics, weak ownership, type erasure, and polymorphic conversions. 

Given the simplicity of our requirements, I proposed a custom implementation using a simple `unsigned int` counter and a `std::mutex`, which I believe is a more appropriate solution here. 

My goal is to use the simplest and most efficient tool for the job. If our requirements involved managing multiple audio devices simultaneously, then the added complexity of `std::shared_ptr` would indeed be justified.

I am open to further discussion and review to ensure the best possible solution for our project.

Reply:
I'm totally okay with this perhaps slightly overpowered use of `std::shared_ptr` and `std::weak_ptr`. @ChrisThrasher has spent a considerable amount of time replacing all the manual memory managements with smart pointers, I'm not about to add our custom MMM just to avoid using a stdlib construct.

Maybe functionality-wise this is "simpler", but it's certainly not simpler implementation wise. Now you have to again ensure that `new` and `delete` are always guaranteed and everything that comes with MMM.

Reply:
> I'm totally okay with this perhaps slightly overpowered use of `std::shared_ptr` and `std::weak_ptr`. @ChrisThrasher has spent a considerable amount of time replacing all the manual memory managements with smart pointers, I'm not about to add our custom MMM just to avoid using a stdlib construct.
> 
> Maybe functionality-wise this is "simpler", but it's certainly not simpler implementation wise. Now you have to again ensure that `new` and `delete` are always guaranteed and everything that comes with MMM.

You raise a great point here -- there's not even any need for dynamic allocation at all in this case. Updated the PR.

Reply:
@eXpl0it3r: After talking to binary, this PR is now ready for your review again.